### PR TITLE
Add `mix hex.registry` build

### PR DESCRIPTION
**Pull request by wojtekmach**
Ref: #776

### CODE REVIEW

File: lib/mix/tasks/hex.registry.ex [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+
+  @shortdoc "Manages local Hex registries"
+
+  @moduledoc """
+  Manages local Hex registries.
+
+  ## Build a local registry
+
+  ```
+  $ mix hex.registry build
```
Comment (wojtekmach) [inline lib/mix/tasks/hex.registry.ex:19]: I wonder if this should be `mix hex.repo build` instead 🤔

File: lib/mix/tasks/hex.registry.ex [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+  $ mix hex.repo add acme http://localhost:8000 --public-key public/public_key
+  $ mix hex.package fetch foo 1.0.0 --repo acme
+  ```
+
+  ### Command line options
+
+    * `--name` - The name of the registry, defaults to the basename of the current working
+      directory
+
+    * `--public-dir` - Path to the directory with public files, defaults to `./public`
```
Comment (ericmj) [inline lib/mix/tasks/hex.registry.ex:55]: I don't think we should provide default. What do you think about passing it as `mix hex.registry build PUBLIC_DIR`.

In fact, maybe the other options should be required as well, but as flags?

File: lib/mix/tasks/hex.registry.ex [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+  * creating public/packages/foo
+  * updating public/names
+  * updating public/versions
+  ```
+
+  You can test the repository by starting the built-in Erlang/OTP HTTP server, adding the repository,
+  and retrieving the package that you just published.
+
+  ```
+  $ erl -s inets -eval 'inets:start(httpd,[{port,8000},{server_name,"localhost"},{server_root,"."},{document_root,"public"}]).'
```
Comment (ericmj) [inline lib/mix/tasks/hex.registry.ex:43]: It's really nice that we can run a simple server without having to install anything 👍

File: lib/mix/tasks/hex.registry.ex [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+        [name, _version] = String.split(Path.basename(path), ["-", ".tar"], trim: true)
+        name
+      end)
+
+    versions =
+      Enum.map(paths_per_name, fn {name, paths} ->
+        releases =
+          for path <- paths do
+            build_release(repo_name, path)
+          end
```
Comment (ericmj) [inline lib/mix/tasks/hex.registry.ex:107]: This will look better when piped with `Enum.map` imo.

File: lib/mix/tasks/hex.registry.ex [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+    build(repo_name, public_dir, private_key)
+  end
+
+  defp build(repo_name, public_dir, private_key) do
+    ensure_public_key(private_key, public_dir)
+    create_directory(Path.join(public_dir, "tarballs"))
+
+    paths_per_name =
+      Path.wildcard("#{public_dir}/tarballs/*.tar")
+      |> Enum.group_by(fn path ->
```
Comment (ericmj) [inline lib/mix/tasks/hex.registry.ex:97]: No pipe needed here.

File: lib/mix/tasks/hex.registry.ex (hunk 2 above)
Comment (wojtekmach) [reply lib/mix/tasks/hex.registry.ex:55]: If private key is required then we no longer automatically generate it, but we could write in the docs for folks to run: `openssl genrsa -out ...` so I could remove a bunch of code. I think —name being optional was kinda neat but yeah, no need, we can always add it later. And if remembering these flags is annoying folks can wrap them in a rebuild.sh script.

Thanks! 👍

Comment (wojtekmach) [reply lib/mix/tasks/hex.registry.ex]: I ended up with providing all 3 as options as it seems more consistent but happy to make PUBLIC_DIR special. WDYT?

File: lib/mix/tasks/hex.registry.ex [hunk 6]
```diff
@@ -43,18 +51,17 @@ defmodule Mix.Tasks.Hex.Registry do
...
   ```
 
   ### Command line options
 
-    * `--name` - The name of the registry, defaults to the basename of the current working
-      directory
+    * `--name` - The name of the registry
 
-    * `--public-dir` - Path to the directory with public files, defaults to `./public`
+    * `--public-dir` - Path to the directory with public files
```
Comment (ericmj) [inline lib/mix/tasks/hex.registry.ex:62]: What do you think about adding this to the argument list instead of having it as a flag?

**Review by ericmj — APPROVED** [review]

**Pull request merged**